src/main/java/org/broadinstitute/barclay/argparser/TaggedArgumentParser.java

```diff
@@ -0,0 +1,333 @@
+                // replace it's companion argument value with the surrogate key to be used later to retrieve the
+                // actual values
+                final String pairKey = getSurrogateKeyForTaggedOption(optionString, userArgList[i+1], tagNameAndValues);
+                finalArgList.add(optionPrefix + optionName);
+                finalArgList.add(pairKey);
+                return 2;  // we consumed two arguments, so advance by two
+
+            } else {
+                // the option appears to be tagged, but we're already at the end of the argument list,
+                // and there is no companion value to use
+                throw new CommandLineException("No value found for tagged argument: " + optionString);
+            }
+        }
+    }
+
+    /**
+     * Attempt to retrieve a
+     * @param putativeSurrogateKey putative key to try to retrieve from the surrogate map
+     * @return tagged option pair for this surrogate, or null if no entry
+     */
+    public Pair<String, String> getTaggedOptionForSurrogate(final String putativeSurrogateKey) {
+        return tagSurrogates.get(putativeSurrogateKey);
+    }
+
+    // See if the opt parser would think this is a short option ("-")
+    private static boolean isShortOptionToken(final String argument) {
+        return argument.startsWith( SHORT_OPTION_PREFIX )
+                && !SHORT_OPTION_PREFIX.equals( argument )
+                && !isLongOptionToken( argument );
+    }
+
+    // See if the opt parser would think this is a long option ("--")
+    private static boolean isLongOptionToken(final String argument) {
+        return argument.startsWith( LONG_OPTION_PREFIX );
+    }
+
+    private String getSurrogateKeyForTaggedOption(
+            final String rawOptionString,   // the raw option string provided by the user, including the prefix and option name
+            final String rawArgumentValue,  // the raw argument value provided by the user
+            final String tagString          // the tag string that has been peeled off of the raw option string, including logical name and any attributes
+    )
+    {
+        final String surrogateKey = makeSurrogateKey(rawOptionString, rawArgumentValue);
+        if (null != tagSurrogates.put(surrogateKey, Pair.of(tagString, rawArgumentValue))) {
+            throw new CommandLineException.BadArgumentValue(
+                    String.format("The argument value: \"%s %s\" was duplicated on the command line", rawOptionString, rawArgumentValue));
+        }
+        return surrogateKey;
+    }
+
+    /**
+     * Construct a surrogate key from a raw option string (as provided by the user) and raw argument value (as
+     * provided by the user). In order to ensure uniqueness of the key, it needs to include all elements from the
+     * command line, including:
+     *
+     *  -the option prefix (short/long) used
```

Comment:
Does the key actually start with the `--`/`-` prefix? I think it's stripped based on my reading of the code, but if it's not stripped it should be sanitized to avoid confusing the parser with an argument value that looks like an argument name.